pkg/grpc/grpc.go

```diff
@@ -202,15 +207,23 @@ func (p *connectionPool) Share(address string) (*grpc.ClientConn, bool) {
 		return nil, false
 	}
 
+	p.referenceLock.Lock()
+	defer p.referenceLock.Unlock()
+
 	p.referenceCount[conn]++
 	return conn, true
 }
 
 func (p *connectionPool) Release(conn *grpc.ClientConn) {
+	p.referenceLock.RLock()
 	if _, ok := p.referenceCount[conn]; !ok {
```

Comment:
Hi, yaron, I would like to know what are the odds of this if happening? 

- if very high: then we should keep this optimization to use a lightweight read lock and expect most of the call be return in this if check.
-  if not very high: then I suggest to remove this optimization to use a BIG lock for the whole method scope. And if we do this, we can replace this RW lock with a simple (and also be quick) lock since we only use READ lock here.

Reply:
Looks like perf is unaffected. cc @skyao

Reply:
@yaron let's quickly competed this PR to make perf test pass.